File: uninformed_search/bfs.py
def bfs(graph, start):
    visited = {}
    for i in graph:
        visited[i] = False
    queue = [start]
    visited[start] = True
    while queue:
        start = queue.pop(0)
        if len(start) > 1:
            start = start[0]
        print(start)
        for i in graph[start]:
            node = i[0]
            if node == "":
                continue
            if not visited[node]:
                queue.append(i)
                visited[node] = True

File: Lab1-AI/src/main/java/com/example/lab1ai/BFS.java
package com.example.lab1ai;

import java.util.*;

public class BFS extends Actions{
    public Stack<State> finalPath = new Stack<>();
    public boolean bfs(String initialState){

        Queue<State> frontiers = new LinkedList<>();
        Set<State> explored = new HashSet<>();


        State INState = new State();
        INState.setBoardState(initialState);
        frontiers.add(INState);

        while (!frontiers.isEmpty()){
            State state = frontiers.remove();
            explored.add(state);
            if (state.isGoal()){
                printPath(INState, state);
                finalPath=getPath(INState, state);
                printExpanded(explored);
                return true;
            }
            expand(state);
            ArrayList<State> neighbors = state.getNeighbors();
            for (State neighbor : neighbors){
                if (!(inFrontiers(frontiers, neighbor)) && !(inExplored(explored, neighbor))){
                    frontiers.add(neighbor);
                }
            }
        }
        System.out.println("Failure");
        return false;
    }

    public static boolean inFrontiers(Queue<State> frontiers, State st){
        return frontiers.stream().map(State::getBoardState).anyMatch(st.getBoardState()::equals);
    }
    //check if state is explored (visited)
    public static boolean inExplored(Set<State> explored, State st){
        return explored.stream().map(State::getBoardState).anyMatch(st.getBoardState()::equals);
    }


}
